In [5]:
import pandas as pd
import numpy as np
from pathlib import Path

In [6]:
nav_df = pd.read_csv("../data/raw/02_nav_history.csv")

print(nav_df.shape)
print(nav_df.head())
print(nav_df.columns)

(46000, 3)
   amfi_code        date      nav
0     119551  2022-01-03  54.3856
1     119551  2022-01-04  54.3474
2     119551  2022-01-05  54.6869
3     119551  2022-01-06  55.4550
4     119551  2022-01-07  55.3692
Index(['amfi_code', 'date', 'nav'], dtype='object')


In [7]:
# Convert date

nav_df["date"] = pd.to_datetime(nav_df["date"])

# Sort

nav_df = nav_df.sort_values(
    ["amfi_code", "date"]
)

# Remove duplicates

duplicates_before = nav_df.duplicated().sum()

nav_df = nav_df.drop_duplicates()

duplicates_after = nav_df.duplicated().sum()

print("Duplicates before:", duplicates_before)
print("Duplicates after:", duplicates_after)

Duplicates before: 0
Duplicates after: 0


In [8]:
invalid_nav = nav_df[
    nav_df["nav"] <= 0
]

print(
    "Invalid NAV rows:",
    len(invalid_nav)
)

Invalid NAV rows: 0


In [9]:
nav_df.to_csv(
    "../data/processed/clean_nav_history.csv",
    index=False
)

print("Saved successfully")

Saved successfully


In [10]:
print(nav_df.columns.tolist())

['amfi_code', 'date', 'nav']


In [11]:
print(nav_df.isnull().sum())


amfi_code    0
date         0
nav          0
dtype: int64


In [12]:
duplicates = nav_df.duplicated().sum()

print("Duplicate rows:", duplicates)


Duplicate rows: 0


In [13]:
invalid_nav = nav_df[nav_df["nav"] <= 0]

print("Invalid NAV rows:", len(invalid_nav))

Invalid NAV rows: 0


In [14]:
nav_df.groupby("amfi_code").size().describe()

count      40.0
mean     1150.0
std         0.0
min      1150.0
25%      1150.0
50%      1150.0
75%      1150.0
max      1150.0
dtype: float64

In [15]:
tx_df = pd.read_csv(
    "../data/raw/08_investor_transactions.csv"
)

print(tx_df.columns.tolist())
print(tx_df.head())

['investor_id', 'transaction_date', 'amfi_code', 'transaction_type', 'amount_inr', 'state', 'city', 'city_tier', 'age_group', 'gender', 'annual_income_lakh', 'payment_mode', 'kyc_status']
  investor_id transaction_date  amfi_code transaction_type  amount_inr  \
0   INV003054       2024-01-01     119092              SIP        1834   
1   INV002952       2024-01-01     148567       Redemption      392882   
2   INV003420       2024-01-01     118636              SIP         912   
3   INV003436       2024-01-01     118634              SIP        1102   
4   INV004691       2024-01-01     119094          Lumpsum        8682   

         state       city city_tier age_group  gender  annual_income_lakh  \
0    Telangana  Hyderabad       T30       56+  Female                77.1   
1       Punjab   Amritsar       B30     18-25    Male                 7.1   
2      Haryana  Faridabad       B30     36-45    Male                47.2   
3  Maharashtra     Mumbai       T30     36-45  Female      

Observation:
Each AMFI scheme contains exactly 1150 NAV observations.
No inconsistencies were found in record counts across schemes.
No missing NAV history was detected at the scheme level.

# investor transaction csv

In [16]:
tx_df["transaction_date"] = pd.to_datetime(
    tx_df["transaction_date"]
)

print(tx_df.dtypes)

investor_id                   object
transaction_date      datetime64[ns]
amfi_code                      int64
transaction_type              object
amount_inr                     int64
state                         object
city                          object
city_tier                     object
age_group                     object
gender                        object
annual_income_lakh           float64
payment_mode                  object
kyc_status                    object
dtype: object


In [17]:
print(
    tx_df["transaction_type"].unique()
)

['SIP' 'Redemption' 'Lumpsum']


In [18]:
invalid_amounts = tx_df[
    tx_df["amount_inr"] <= 0
]

print(
    "Invalid Amount Rows:",
    len(invalid_amounts)
)

Invalid Amount Rows: 0


In [19]:
print(
    tx_df["kyc_status"].unique()
)

['Verified' 'Pending']


In [20]:
print(
    tx_df["payment_mode"].unique()
)

['UPI' 'Cheque' 'Mandate' 'Net Banking']


In [21]:
print(
    tx_df.isnull().sum()
)

investor_id           0
transaction_date      0
amfi_code             0
transaction_type      0
amount_inr            0
state                 0
city                  0
city_tier             0
age_group             0
gender                0
annual_income_lakh    0
payment_mode          0
kyc_status            0
dtype: int64


In [22]:
duplicates = tx_df.duplicated().sum()

print(
    "Duplicate Rows:",
    duplicates
)

Duplicate Rows: 0


In [23]:
tx_df.to_csv(
    "../data/processed/clean_investor_transactions.csv",
    index=False
)

# 07_scheme_performance.csv

In [24]:
perf_df = pd.read_csv(
    "../data/raw/07_scheme_performance.csv"
)

print(perf_df.columns.tolist())
print(perf_df.head())

['amfi_code', 'scheme_name', 'fund_house', 'category', 'plan', 'return_1yr_pct', 'return_3yr_pct', 'return_5yr_pct', 'benchmark_3yr_pct', 'alpha', 'beta', 'sharpe_ratio', 'sortino_ratio', 'std_dev_ann_pct', 'max_drawdown_pct', 'aum_crore', 'expense_ratio_pct', 'morningstar_rating', 'risk_grade']
   amfi_code                                   scheme_name       fund_house  \
0     119551     SBI Bluechip Fund - Regular Plan - Growth  SBI Mutual Fund   
1     119552      SBI Bluechip Fund - Direct Plan - Growth  SBI Mutual Fund   
2     119598    SBI Small Cap Fund - Regular Plan - Growth  SBI Mutual Fund   
3     119599     SBI Small Cap Fund - Direct Plan - Growth  SBI Mutual Fund   
4     119120  SBI Magnum Gilt Fund - Regular Plan - Growth  SBI Mutual Fund   

    category     plan  return_1yr_pct  return_3yr_pct  return_5yr_pct  \
0  Large Cap  Regular           12.42           12.36           14.45   
1  Large Cap   Direct           15.25           11.30           14.23   
2  Small 

In [25]:
print(perf_df.isnull().sum())

amfi_code             0
scheme_name           0
fund_house            0
category              0
plan                  0
return_1yr_pct        0
return_3yr_pct        0
return_5yr_pct        0
benchmark_3yr_pct     0
alpha                 0
beta                  0
sharpe_ratio          0
sortino_ratio         0
std_dev_ann_pct       0
max_drawdown_pct      0
aum_crore             0
expense_ratio_pct     0
morningstar_rating    0
risk_grade            0
dtype: int64


In [27]:
numeric_cols = [
    "return_1yr_pct",
    "return_3yr_pct",
    "return_5yr_pct",
    "benchmark_3yr_pct",
    "alpha",
    "beta",
    "sharpe_ratio",
    "sortino_ratio",
    "std_dev_ann_pct",
    "max_drawdown_pct",
    "expense_ratio_pct"
]

print(perf_df[numeric_cols].dtypes)

return_1yr_pct       float64
return_3yr_pct       float64
return_5yr_pct       float64
benchmark_3yr_pct    float64
alpha                float64
beta                 float64
sharpe_ratio         float64
sortino_ratio        float64
std_dev_ann_pct      float64
max_drawdown_pct     float64
expense_ratio_pct    float64
dtype: object


In [28]:
invalid_expense = perf_df[
    (perf_df["expense_ratio_pct"] < 0.1) |
    (perf_df["expense_ratio_pct"] > 2.5)
]

print(
    "Invalid Expense Ratios:",
    len(invalid_expense)
)

Invalid Expense Ratios: 0


In [29]:
negative_sharpe = perf_df[
    perf_df["sharpe_ratio"] < 0
]

print(
    "Negative Sharpe Funds:",
    len(negative_sharpe)
)

Negative Sharpe Funds: 0


In [31]:
duplicates = perf_df.duplicated().sum()

print(
    "Duplicate Rows:",
    duplicates
)

Duplicate Rows: 0


In [32]:
perf_df.to_csv(
    "../data/processed/clean_scheme_performance.csv",
    index=False
)

In [33]:
print(tx_df["transaction_type"].unique())

print(tx_df["kyc_status"].unique())

print(tx_df.duplicated().sum())

print(tx_df.isnull().sum())

print(perf_df.isnull().sum())

print("Invalid Expense Ratios:", len(invalid_expense))

print("Negative Sharpe Funds:", len(negative_sharpe))

['SIP' 'Redemption' 'Lumpsum']
['Verified' 'Pending']
0
investor_id           0
transaction_date      0
amfi_code             0
transaction_type      0
amount_inr            0
state                 0
city                  0
city_tier             0
age_group             0
gender                0
annual_income_lakh    0
payment_mode          0
kyc_status            0
dtype: int64
amfi_code             0
scheme_name           0
fund_house            0
category              0
plan                  0
return_1yr_pct        0
return_3yr_pct        0
return_5yr_pct        0
benchmark_3yr_pct     0
alpha                 0
beta                  0
sharpe_ratio          0
sortino_ratio         0
std_dev_ann_pct       0
max_drawdown_pct      0
aum_crore             0
expense_ratio_pct     0
morningstar_rating    0
risk_grade            0
dtype: int64
Invalid Expense Ratios: 0
Negative Sharpe Funds: 0


In [34]:
import pandas as pd

fund_df = pd.read_csv("../data/raw/01_fund_master.csv")
aum_df = pd.read_csv("../data/raw/03_aum_by_fund_house.csv")

print(fund_df.shape)
print(aum_df.shape)

(40, 15)
(90, 5)


In [35]:
from pathlib import Path

processed = Path("../data/processed")

for file in processed.glob("*.csv"):
    print(file.name)


clean_investor_transactions.csv
clean_nav_history.csv
clean_scheme_performance.csv


In [36]:
import pandas as pd
from pathlib import Path

raw_dir = Path("../data/raw")
processed_dir = Path("../data/processed")

# 1. Fund Master
fund_master = pd.read_csv(raw_dir / "01_fund_master.csv")
fund_master.to_csv(
    processed_dir / "clean_fund_master.csv",
    index=False
)

# 2. AUM By Fund House
aum_df = pd.read_csv(raw_dir / "03_aum_by_fund_house.csv")

aum_df["date"] = pd.to_datetime(aum_df["date"])

aum_df.to_csv(
    processed_dir / "clean_aum_by_fund_house.csv",
    index=False
)

# 3. Monthly SIP Inflows
sip_df = pd.read_csv(raw_dir / "04_monthly_sip_inflows.csv")

sip_df["month"] = pd.to_datetime(
    sip_df["month"]
)

sip_df.to_csv(
    processed_dir / "clean_monthly_sip_inflows.csv",
    index=False
)

# 4. Category Inflows
cat_df = pd.read_csv(raw_dir / "05_category_inflows.csv")

cat_df["month"] = pd.to_datetime(
    cat_df["month"]
)

cat_df.to_csv(
    processed_dir / "clean_category_inflows.csv",
    index=False
)

# 5. Industry Folio Count
folio_df = pd.read_csv(
    raw_dir / "06_industry_folio_count.csv"
)

folio_df["month"] = pd.to_datetime(
    folio_df["month"]
)

folio_df.to_csv(
    processed_dir / "clean_industry_folio_count.csv",
    index=False
)

# 6. Portfolio Holdings
portfolio_df = pd.read_csv(
    raw_dir / "09_portfolio_holdings.csv"
)

if "portfolio_date" in portfolio_df.columns:
    portfolio_df["portfolio_date"] = pd.to_datetime(
        portfolio_df["portfolio_date"]
    )

portfolio_df.to_csv(
    processed_dir / "clean_portfolio_holdings.csv",
    index=False
)

# 7. Benchmark Indices
benchmark_df = pd.read_csv(
    raw_dir / "10_benchmark_indices.csv"
)

benchmark_df["date"] = pd.to_datetime(
    benchmark_df["date"]
)

benchmark_df.to_csv(
    processed_dir / "clean_benchmark_indices.csv",
    index=False
)

print("All remaining datasets cleaned and saved successfully.")

All remaining datasets cleaned and saved successfully.


In [37]:
from pathlib import Path

processed = Path("../data/processed")

for file in sorted(processed.glob("*.csv")):
    print(file.name)

clean_aum_by_fund_house.csv
clean_benchmark_indices.csv
clean_category_inflows.csv
clean_fund_master.csv
clean_industry_folio_count.csv
clean_investor_transactions.csv
clean_monthly_sip_inflows.csv
clean_nav_history.csv
clean_portfolio_holdings.csv
clean_scheme_performance.csv


In [39]:
from sqlalchemy import create_engine

engine = create_engine(
    "sqlite:///../data/db/bluestock_mf.db"
)

print("Database created successfully")

Database created successfully


In [40]:
from pathlib import Path

db_path = Path("../data/db/bluestock_mf.db")

print(db_path.resolve())

C:\Users\purua\projects\bluestock_mf_capstone\data\db\bluestock_mf.db


In [41]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(
    "sqlite:///../data/db/bluestock_mf.db"
)

test_df = pd.DataFrame({
    "id": [1]
})

test_df.to_sql(
    "test_table",
    engine,
    if_exists="replace",
    index=False
)

print("Database created")

Database created


In [42]:
fund_df = pd.read_csv(
    "../data/processed/clean_fund_master.csv"
)

fund_df.to_sql(
    "dim_fund",
    engine,
    if_exists="replace",
    index=False
)

print("dim_fund loaded")

dim_fund loaded


In [43]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(
    "sqlite:///../data/db/bluestock_mf.db"
)

tables = {
    "dim_fund":
        "../data/processed/clean_fund_master.csv",

    "fact_nav":
        "../data/processed/clean_nav_history.csv",

    "fact_transactions":
        "../data/processed/clean_investor_transactions.csv",

    "fact_performance":
        "../data/processed/clean_scheme_performance.csv",

    "fact_aum":
        "../data/processed/clean_aum_by_fund_house.csv"
}

for table_name, file_path in tables.items():

    df = pd.read_csv(file_path)

    df.to_sql(
        table_name,
        engine,
        if_exists="replace",
        index=False
    )

    print(
        f"{table_name}: {len(df)} rows loaded"
    )

dim_fund: 40 rows loaded
fact_nav: 46000 rows loaded
fact_transactions: 32778 rows loaded
fact_performance: 40 rows loaded
fact_aum: 90 rows loaded


In [44]:
tables = pd.read_sql(
    "SELECT name FROM sqlite_master WHERE type='table';",
    engine
)

print(tables)

                name
0         test_table
1           dim_fund
2           fact_nav
3  fact_transactions
4   fact_performance
5           fact_aum


In [45]:
for table in [
    "dim_fund",
    "fact_nav",
    "fact_transactions",
    "fact_performance",
    "fact_aum"
]:
    count = pd.read_sql(
        f"SELECT COUNT(*) AS rows FROM {table}",
        engine
    )

    print(table)
    print(count)

dim_fund
   rows
0    40
fact_nav
    rows
0  46000
fact_transactions
    rows
0  32778
fact_performance
   rows
0    40
fact_aum
   rows
0    90


In [46]:
for table in [
    "dim_fund",
    "fact_nav",
    "fact_transactions",
    "fact_performance",
    "fact_aum"
]:
    count = pd.read_sql(
        f"SELECT COUNT(*) AS rows FROM {table}",
        engine
    )

    print(table)
    print(count)
    print("-"*50)

dim_fund
   rows
0    40
--------------------------------------------------
fact_nav
    rows
0  46000
--------------------------------------------------
fact_transactions
    rows
0  32778
--------------------------------------------------
fact_performance
   rows
0    40
--------------------------------------------------
fact_aum
   rows
0    90
--------------------------------------------------
